### ISSUE TITLE
Client can't startup when backend is down.

### ISSUE THREAD

**Issue by sjcjonker** [cablelabs/dis-arbor-monitor#10]
During init the client is creating a new DIS Client instance at [line 491](https://github.com/cablelabs/dis-arbor-monitor/blob/1e58cb3774d6e708b91241af0c805efcaf8256cd/arbor-monitor/__main__.py#L491). This causes the SDK to call
`self._dest_url = f"{self._base_url}/data?api_key={self._key}"`

This results in an exception:
```
Which then is resulting in a webcall to the backend resulting itself in:
2021-02-04 20:39:36,881 arborreport: INFO Report storage format: all-attributes
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/app/arbor-monitor/__main__.py", line 492, in <module>
    dis_client = DisClient(api_key=args.report_consumer_api_key)
  File "/app/arbor-monitor/dis_client_sdk/DisClient.py", line 45, in __init__
    raise Exception(res.json())
  File "/usr/local/lib/python3.6/site-packages/requests/models.py", line 900, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/local/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

As the backend returns:
```
[0] curl  https://api.dissarm.net/v1/data\?api_key\=<<REMOVED>>
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.18.0 (Ubuntu)</center>
</body>
</html>
```

**Comment by craigpratt** [comment]
Proposed fix for this issue is on branch task/fix-issue-10

Here's the PR for review: https://github.com/cablelabs/dis-arbor-monitor/pull/11

From the commit:

Changed DisClient constructor to not retrieve info on construct

The client info isn't retrieved until DisClient.get_info()
 is called. And this now throws an exception on anything but
 a 200 status code.

Also made the dis_arbor_monitor main routine catch any
 exception on a call to DisClient.get_info() and log
 it as a warn instead of bailing out.

**Comment by craigpratt** [comment]
@sjcjonker Please let me know if you need a Docker image with this fix and I'll be happy to provide one

**Comment by sjcjonker** [comment]
@craigpratt no that's fine. I'll deploy (git clone, build container etc) this later today and test and get back to you asap.

**Comment by sjcjonker** [comment]
Hi @craigpratt,

Thanks the client does start up. There is only one thing I can't get to work and this is the writing of the JSON files to disk. That said it's a bit hard to troubleshoot now with the backend returning mostly 502's (fyi sometimes it can upload reports as it seems).

What is happening in my view:
- `report_store_dir` is checked and then the validated path is put in `report_storage_path`

Then in [the report upload section](https://github.com/cablelabs/dis-arbor-monitor/blob/43989b85ea4c835e17c576f56aae8a0112a61906/arbor-monitor/__main__.py#L81) line 81 t/m 83, the actual block where the `report_storage_path` is checked on [line 90](https://github.com/cablelabs/dis-arbor-monitor/blob/43989b85ea4c835e17c576f56aae8a0112a61906/arbor-monitor/__main__.py#L90) never gets checked if the upload fails.

I'm therefore concluding if the upload would succeed, only the last file is written. (As said assumption, but can't validate at present I'm afraid).

I do believe we should always write the file, whether the upload succeeds or not.

**Comment by craigpratt** [comment]
I agree - it should definitely save the report regardless.

I just made a change to support this - will push it in a minute and test it on our test network using the mock netscout setup.

**Comment by craigpratt** [comment]
Committed additional change to branch (commit 7be5f2da7d)

**Comment by sjcjonker** [comment]
I'm fighting for the past 15 minutes issues between my fork, your fix-issue-10 branch etc etc etc. So I can't make a pull request. Sorry. Here is the patch I'm running with currently:

```
--- __main__.py 2021-02-05 20:51:42.470611850 +0000
+++ __main__.py.orig    2021-02-05 20:52:10.554838738 +0000
@@ -78,16 +78,15 @@
     if args.dry_run:
         logger.info(f"Attack ID {attack_id}: Running in DRY RUN mode - not posting attack")
     else:
-        upload_or_write_error = False
         try:
             source_ip_list = send_report_to_dis_server(attack_id, payload, src_traffic_report)
-            total_reports_sent += 1
-            total_source_ips_reported += len(source_ip_list)
         except Exception as ex:
             msg = f"Caught an exception uploading the report for attack {attack_id} ({ex})"
             logger.warning(msg)
-            upload_or_write_error = True
+            return jsonify({"warning": msg}), 200, {'Content-Type': 'application/json'}

+        total_reports_sent += 1
+        total_source_ips_reported += len(source_ip_list)
         if report_storage_path:
             try:
                 save_attack_report_file(report_storage_path, args.report_store_format,
@@ -95,14 +94,11 @@
             except Exception as ex:
                 msg = f"Caught an exception saving the report for attack {attack_id} ({ex}) - report uploaded, CONTINUING"
                 logger.warning(msg)
-                upload_or_write_error = True
                 # Not returning a 400 here so Netscout doesn't keep attempting to redeliver this attack
                 #  notification (causing potential duplicate reports and backing up Netscout's notify queue)
+                return jsonify({"warning": msg}), 200, {'Content-Type': 'application/json'}

-        if upload_or_write_error:
-            return jsonify({"warning": msg}), 200, {'Content-Type': 'application/json'}
-        else:
-            return f"Thank you Netscout for the DOS report! (attack ID {attack_id})", 200, {'Content-Type': 'text/plain'}
+    return f"Thank you Netscout for the DOS report! (attack ID {attack_id})", 200, {'Content-Type': 'text/plain'}

 def check_sightline_api_supported():
     """
```

**Comment by craigpratt** [comment]
ah yeah - looks like you were diffing against the change I just made!

git is a bit of a beast. I've been using it for years and still consider myself "semi-proficient at best". There are definitely some good tricks to make life easier which I'm happy to share. But if you just want to make PRs and let me sort it out, that's fine too.

**Comment by sjcjonker** [comment]
> ah yeah - looks like you were diffing against the change I just made!
Sorry yes I downloaded a fresh copy and I think with your changes as well. Just ignore.
Your code does the same, except I added a boolean to track the error.
As per the other comment, I'm running upto commit 7be5f2d now, and it seems fine.
Lets hope the backend gets back online soon, then we can see if the happy scenario still works or we broke it.
